csrc/C++20/ranges

```diff
@@ -0,0 +1,78 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <type_traits>
+
+// This is backport of C++20 ranges library to C++17.
+
+namespace std {
+
+namespace ranges {
+
+template <typename W, typename Bound>
+class iota_view {
+  static_assert(
+      std::is_same_v<std::decay_t<W>, std::decay_t<Bound>>,
+      "W != Bound not supported yet");
+
+ public:
+  constexpr iota_view(W&& value, Bound&& bound)
+      : value_(std::forward<W>(value)), bound_(std::forward<Bound>(bound)) {}
```

Comment:
Good idea to not support the single argument constructor, since that could lead to infinite loops when blindly transitioning from `c10::irange`. I suppose we should just get used to always specifying the start value.

Reply:
Another possibility may be to introduce our own `irange` which uses this backport and has the single argument constructor. That way we can sidestep the need to specify the correct literal zero at each call site.